webui/react/src/hooks/useSettings.ts

```diff
@@ -290,38 +282,57 @@ const useSettings = <T>(config: SettingsConfig<T>): UseSettingsReturn<T> => {
   );
 
   const updateSettings = useCallback(
-    async (updates: Settings, shouldPush = false) => {
-      if (!settings || shouldSkipUpdates) return;
+    (updates: Settings, shouldPushUpdate = false) => {
+      if (shouldSkipUpdates) return;
 
-      const newSettings = { ...settings, ...updates };
+      update(config.storagePath, (settings) => {
+        if (!settings) return updates;
+        return { ...settings, ...updates };
+      });
 
-      if (isEqual(newSettings, settings)) return;
+      setShouldPush(shouldPushUpdate);
+    },
+    [config, update, shouldSkipUpdates],
+  );
 
-      update(config.storagePath, newSettings);
+  useEffect(() => {
+    // updates, if necessary, the returned settings
+    if (!settings || isEqual(settings, returnedSettings)) return;
 
-      await updateDB(newSettings);
+    setReturnedSettings(settings);
 
-      if (
-        (Object.values(config.settings) as SettingsConfigProp<typeof config>[]).every(
-          (setting) => !!setting.skipUrlEncoding,
-        )
-      ) {
-        return;
-      }
+    updateDB(settings); // no need to await for DB changes as we optimistically update stuff based on the provided values
+  }, [settings, returnedSettings, updateDB]);
 
-      const mappedSettings = settingsToQuery(config, newSettings);
-      const url = `?${mappedSettings}`;
+  useEffect(() => {
+    // updates the query settings
+    if (shouldSkipUpdates) return;
 
-      shouldPush ? navigate(url) : navigate(url, { replace: true });
-    },
-    [config, settings, navigate, update, updateDB, shouldSkipUpdates],
-  );
+    if (
+      (Object.values(config.settings) as SettingsConfigProp<typeof config>[]).every(
+        (setting) => !!setting.skipUrlEncoding,
+      )
+    ) {
+      return;
+    }
+
+    const mappedSettings = settingsToQuery(config, returnedSettings);
+    const url = `?${mappedSettings}`;
+
+    if (mappedSettings && location.search !== url) {
+      if (shouldPush) {
+        navigate(url);
+      } else {
+        navigate(url, { replace: true });
+      }
```

Comment:
As Emily commented, 
```suggestion
      navigate(url, { replace: !shouldPush });
```

is more concise